docs/utils/data.rst

```diff
@@ -40,10 +40,10 @@ find the cache has corrupted or outdated data in it, you can remove an entry or
 clear the whole thing with `~astropy.utils.data.clear_download_cache`.
 
 The files in the cache directory are named according to a cryptographic hash of
-their contents (currently MD5, so in principle malevolent entities can cause
-collisions, though the security risks this poses are marginal at most). Thus
-files with the same content share storage. The modification times on these
-files normally indicate when they were last downloaded from the Internet.
+their URL (currently MD5, so in principle malevolent entities can cause
```

Comment:
I can't remember anymore: Before the 4.0 refactoring, was it hashed using contents or URL? In other words, is this something new or is this reverting to 3.x hashing behavior?

If this is something new, I wonder why we were hashing by content before (and how does doing this fix the locking problem). For instance, we might want to make sure some older code relying on `data.py` to provide mirror downloads still work as expected. A quick glance at `master` code suggests maybe this is already addressed but just want to double check.

Reply:
Hashing by contents provided deduplication. But it is necessary to store the URL somewhere, and because the old system stored the URL information in a separate shelve object, locking was necessary. Basing the filename on the URL lets the filesystem itself serve as the lookup index, so the concurrency handling can rely on the operating system.